src/circuit/num.rs

```diff
@@ -696,6 +696,48 @@ impl<E: Engine> AllocatedNum<E> {
         Ok(())
     }
 
+    /// Takes two allocated numbers (a, b) and returns
+    /// allocated boolean variable with value `true`
+    /// if the `a` smaller than `b`, `false` otherwise.
+    pub fn smaller<CS>(
+        mut cs: CS,
+        a: &Self,
+        b: &Self
+    ) -> Result<boolean::AllocatedBit, SynthesisError>
+        where E: Engine,
+            CS: ConstraintSystem<E>
+    {
+        // Let `delta = a - b`
+
+        let delta_value = match (a.value, b.value) {
+            (Some(a), Some(b))  => {
+                // return (a - b)
+                let mut a = a;
+                a.sub_assign(&b);
+                Some(a)
+            },
+            _ => None,
+        };
+
+        let delta = Self::alloc(cs.namespace(|| "delta"), || delta_value.grab())?;
+
+        /// delta equal to a - b
+        cs.enforce(
+            || "delta equality",
+            |lc| lc + delta.get_variable(),
+            |lc| lc + CS::one(),
+            |lc| lc + a.get_variable() - b.get_variable()
+        );
+
+        /// Let 'bits' - 255 bits representation of delta
+        let bits = delta.into_bits_le(&mut cs)?;
+
+        match bits[254].get_variable() {
```

Comment:
254 is not universal. Next test panics:

```rust
    #[test]
    fn test_less_than() -> Result<(), SynthesisError> {
        use bellman::pairing::bn256::{Bn256, Fr};

        let mut cs = TestConstraintSystem::<Bn256>::new();

        let a = AllocatedNum::alloc(cs.namespace(|| "a"), || Ok(Fr::from_str("10").unwrap()))?;
        let b = AllocatedNum::alloc(cs.namespace(|| "b"), || Ok(Fr::from_str("12").unwrap()))?;

        let num = AllocatedNum::smaller(cs.namespace(|| "a < b"), &a, &b)?;
        assert_eq!(num.get_value(), Some(true));

        Ok(())
    }
```

Reply:
Such check is proper for uintX arithmetic (where X is strictly less than bit length of the field), but I’m not sure about direct field elements. E.g for a prime 13: 0 - 12 = 1 and will have zero in the top bit